ENH support float32 in SpectralEmbedding for LOBPCG and PyAMG solvers

lobpcg opened the pull request:
Remove check_array call for float64, since both lobpcg function and pyamg package support float32 for a few years now; see
https://github.com/scikit-learn/scikit-learn/issues/21321


#### Reference Issues/PRs
Fixes #21321.

#### What does this implement/fix? Explain your changes.
Remove check_array call for float64, since both lobpcg function and pyamg package support float32 for a few years now

lobpcg committed:
Update _spectral_embedding.py

Remove check_array call for float64, since both lobpcg function and pyamg package support float32 for a few years now; see
https://github.com/scikit-learn/scikit-learn/issues/21321

lobpcg committed:
Update test_spectral_embedding.py

test_spectral_embedding_two_components for dtype in [np.float32, np.float64] and @pytest.mark.parametrize("eigen_solver", ("arpack", "lobpcg", "amg"))

lobpcg committed:
Update test_spectral_embedding.py

typo

lobpcg committed:
Update test_spectral_embedding.py

mistake fixed

lobpcg committed:
black

lobpcg committed:
mistake fixed

lobpcg committed:
added `amgloaded` distinction

lobpcg committed:
lint

lobpcg committed:
black --line-length 79 plus pytest.warns -> pytest.raises

lobpcg committed:
pass + black

lobpcg committed:
Update v1.1.rst

added an entry

lobpcg committed:
Update v1.1.rst

PR 21534 entered

lobpcg committed:
@pytest.mark.parametrize("eigen_solver", ("arpack", "lobpcg", "amg")) @pytest.mark.parametrize("dtype", ("np.float32", "np.float64")) added to test_spectral_embedding_precomputed_affinity

lobpcg committed:
Merge branch 'patch-1' of https://github.com/lobpcg/scikit-learn into patch-1

lobpcg committed:
np.float32 ->'float32' etc plus amg tests

lobpcg commented:
It's not ready to merge, but ready for a first round of review. 
The change in the function is trivial. All the issues are with unit tests, where I could use a bit of help/advice. 

There is an error that I need help with please to understand and fix:
```
2021-11-03T01:37:44.4112504Z >               rbf_kernel(X.astype(dtype), gamma=gamma)
2021-11-03T01:37:44.4113255Z             )
2021-11-03T01:37:44.4113514Z 
2021-11-03T01:37:44.4114158Z X          = <1000x5 sparse matrix of type '<class 'numpy.float64'>'
2021-11-03T01:37:44.4115211Z 	with 5000 stored elements in Compressed Sparse Row format>
2021-11-03T01:37:44.4115908Z dtype      = 'np.float32'
2021-11-03T01:37:44.4116614Z eigen_solver = 'lobpcg'
2021-11-03T01:37:44.4117099Z gamma      = 1.0
2021-11-03T01:37:44.4118246Z se_precomp = SpectralEmbedding(affinity='precomputed', eigen_solver='lobpcg',
2021-11-03T01:37:44.4119329Z                   random_state=RandomState(MT19937) at 0x7F6B0FC36940)
2021-11-03T01:37:44.4120391Z se_rbf     = SpectralEmbedding(affinity='rbf', eigen_solver='lobpcg', gamma=1.0,
2021-11-03T01:37:44.4121035Z                   random_state=RandomState(MT19937) at 0x7F6B0FC36740)
2021-11-03T01:37:44.4121705Z seed       = 36
2021-11-03T01:37:44.4121979Z 
2021-11-03T01:37:44.4122665Z [1m[31m../1/s/sklearn/manifold/tests/test_spectral_embedding.py[0m:182: 
2021-11-03T01:37:44.4123485Z _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
2021-11-03T01:37:44.4123856Z 
2021-11-03T01:37:44.4124469Z self = <1000x5 sparse matrix of type '<class 'numpy.float64'>'
2021-11-03T01:37:44.4125058Z 	with 5000 stored elements in Compressed Sparse Row format>
2021-11-03T01:37:44.4125773Z dtype = 'np.float32', casting = 'unsafe', copy = True
2021-11-03T01:37:44.4126125Z 
2021-11-03T01:37:44.4129262Z     [34;01mdef[39;49;00m [32;01mastype[39;49;00m([36;01mself[39;49;00m, dtype, casting=[33m'[39;49;00m[33munsafe[39;49;00m[33m'[39;49;00m, copy=[36;01mTrue[39;49;00m):
2021-11-03T01:37:44.4130535Z >       dtype = np.dtype(dtype)
2021-11-03T01:37:44.4131349Z [1m[31mE       TypeError: data type "np.float32" not understood[0m
2021-11-03T01:37:44.4131791Z 
2021-11-03T01:37:44.4132388Z casting    = 'unsafe'
2021-11-03T01:37:44.4133344Z copy       = True
2021-11-03T01:37:44.4134140Z dtype      = 'np.float32'
2021-11-03T01:37:44.4135328Z self       = <1000x5 sparse matrix of type '<class 'numpy.float64'>'
2021-11-03T01:37:44.4135974Z 	with 5000 stored elements in Compressed Sparse Row format>
```

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -258,12 +300,14 @@ def test_spectral_embedding_amg_solver(seed=36):
     "ignore:scipy.linalg.pinv2 is deprecated:DeprecationWarning:pyamg.*"
 )
-def test_spectral_embedding_amg_solver_failure():
+@pytest.mark.parametrize("dtype", ("'float32'", "'float64'"))
```
This should solve the `astype` error

```suggestion
@pytest.mark.parametrize("dtype", (np.float32, np.float64))
```

lobpcg replied:
I have tried it first - gives an error. Now testing
@pytest.mark.parametrize("dtype", ("float32'", "'float64"))

lobpcg committed:
Update sklearn/manifold/tests/test_spectral_embedding.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

ogrisel commented on sklearn/manifold/_spectral_embedding.py:
```diff
@@ -315,8 +315,6 @@ def spectral_embedding(
         if not sparse.issparse(laplacian):
             warnings.warn("AMG works better for sparse matrices")
-        # lobpcg needs double precision floats
-        laplacian = check_array(laplacian, dtype=np.float64, accept_sparse=True)
```
we still want to check that the inputs are valid on public API.

```python
laplacian = check_array(laplacian, dtype=[np.float64, np.float32], accept_sparse=True)
```

lobpcg replied:
make sense, indeed, thanks!

ogrisel commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -117,33 +126,64 @@ def test_spectral_embedding_two_components(seed=36):
+            # ):
+            #     embedded_coordinate = se_precomp.fit_transform(
+            #         affinity.astype(dtype)
+            #     )
```
Why is this commented out?

lobpcg replied:
I was trying to check for the `pyamg not installed` error here, but this would not pass pytest, so there was something wrong in the way these lines were written. Could you please advice how to do it correctly?

lobpcg replied:
This appears to be the last issue to resolve

ogrisel replied:
Maybe this would help?

https://github.com/scikit-learn/scikit-learn/pull/21534/files#r742771584

In any case, feel free to let it fail on the CI rather than pushing the commented out version so that everybody can see the error message without having to pull the branch and run the test locally.

lobpcg replied:
Thanks, fixed, see https://github.com/scikit-learn/scikit-learn/pull/21534/files#r742915773

ogrisel commented on sklearn/manifold/_spectral_embedding.py:
```diff
@@ -346,8 +344,6 @@ def spectral_embedding(
 
     if eigen_solver == "lobpcg":
-        # lobpcg needs double precision floats
-        laplacian = check_array(laplacian, dtype=np.float64, accept_sparse=True)
```
```python
laplacian = check_array(laplacian, dtype=[np.float64, np.float32], accept_sparse=True)
```

ogrisel reviewed (commented):
It's still needed to check the input values but we can change the checks to accept float32 data:

lobpcg committed:
Update _spectral_embedding.py

laplacian = check_array(laplacian, dtype=[np.float64, np.float32], accept_sparse=True)

lobpcg committed:
Update test_spectral_embedding.py

trying everywhere:
@pytest.mark.parametrize("dtype", (float32, float64))

lobpcg committed:
Update test_spectral_embedding.py

@pytest.mark.parametrize("dtype", ("float32", "float64"))

lobpcg committed:
Update _spectral_embedding.py

formatting
        laplacian = check_array(
            laplacian,
            dtype=[np.float64, np.float32],
            accept_sparse=True
        )

lobpcg committed:
Update _spectral_embedding.py

black

lobpcg commented:
black fails here, but passes on my local system. I have not seen this before.

There is no way to fix it for me, since black here does not show the changes suggested.

lobpcg committed:
Update test_spectral_embedding.py

@pytest.mark.parametrize("dtype", ("float32", "float64")) works, but now try @pytest.mark.parametrize("dtype", (np.float32, np.float64))

glemaitre commented:
Regarding `black`, you need to make sure to have install the specific version: `black==21.6b0`.
A nice workaround to note wonder too much it to do the optional install of `pre-commit` described in https://scikit-learn.org/dev/developers/contributing.html#how-to-contribute (item 9.)

Then, at the first `git commit -am"my commit"`, `pre-commit` will run and change all the files using black (and optionally give the failure with `mypy` and `flake8` that will be provided by the CI) if necessary. This could be helpful.

glemaitre commented on sklearn/manifold/_spectral_embedding.py:
```diff
@@ -632,8 +634,7 @@ def fit(self, X, y=None):
                     "'precomputed', 'rbf', 'nearest_neighbors' "
-                    "or a callable."
-                    % self.affinity
+                    "or a callable." % self.affinity
```
This is the change that I have if I run the specific `black` version.

```suggestion
                    "or a callable."
                    % self.affinity
```

lobpcg committed:
Update sklearn/manifold/_spectral_embedding.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

lobpcg commented:
> Regarding `black`, you need to make sure to have install the specific version: `black==21.6b0`. 

Just removed my 21.10b and installed specifically `py -3 -m pip install black==21.6b0` still no change in the py file locally. I run all this in MS Win, but that should probably not matter. Strange, but anyway, thanks very much for your help with the proposed edit! 

May be when black runs in CI, it could actually show the proposed changes - why does it run silently now?

glemaitre commented:
> May be when black runs in CI, it could actually show the proposed changes - why does it run silently now?

Yep, it could be nice to show the diff. I think that `black` should have a `--diff` option. We will make a PR. Thanks for the suggestion.

lobpcg commented:
I have found another issue that needs attention: 
```
#initial approximation X to eigenvectors
X = random_state.rand(laplacian.shape[0], n_components + 1)
```
must in fact be Gaussian (while I guess `rand` is uniform) and should better match the dtype of the matrix.
I'll change it next

lobpcg committed:
Update _spectral_embedding.py

initial X rand -> randn and matched tdype, for lobpcg and amg

lobpcg commented:
I have found another stalled issue with this same function, https://github.com/scikit-learn/scikit-learn/pull/11968 
Should probably be a separate PR. I can try to do it after this PR is merged while my memory is fresh.

ogrisel commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -117,33 +126,64 @@ def test_spectral_embedding_two_components(seed=36):
+        #     embedded_coordinate = se_precomp.fit_transform(
+        #         affinity.astype(dtype)
+        #     )
+        pass
```
Either we need to enable this check for the `ValueError` and its message or just skip this test with `pytest.importorskip("pyamg")`.

Note that sometimes it's necessary to wrap the expected error message with `re.escape`, e.g.:

```python
match=re.escape("The eigen_solver was set to 'amg', but")
```

since the `match` parameter expects a regular expression from the `re` module.

lobpcg replied:
Thanks!
I have found another way at https://code-maven.com/pytest-assert-error-reporting that separates assertions of  raising the error and the content  of the error message, which is nicer. Testing it now.

ogrisel commented on sklearn/manifold/_spectral_embedding.py:
```diff
@@ -364,8 +367,9 @@ def spectral_embedding(
             # doesn't behave well in low dimension and create initial
             # approximation X to eigenvectors
-            X = random_state.rand(laplacian.shape[0], n_components + 1)
+            X = random_state.randn(laplacian.shape[0], n_components + 1)
```
Ideally this would have been fixed in a separate PR with its own changelog entry and potentially a new non-regression test (if possible).

If it's too complicated to disentangle the two fixes, then at least please add a new changelog entry to document this fix that also impacts the solver when running with `np.float64` data.

lobpcg replied:
not complicated to disentangle the two fixes, so I remove it from here and make a separate PR after this one is merged

lobpcg replied:
I have created a separate issue https://github.com/scikit-learn/scikit-learn/issues/21555 to address this in a separate PR

ogrisel commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -205,7 +245,9 @@ def test_spectral_embedding_callable_affinity(X, seed=36):
-def test_spectral_embedding_amg_solver(seed=36):
+@pytest.mark.parametrize("dtype", (np.float32, np.float64))
+def test_spectral_embedding_amg_solver(dtype):
+    seed = 36
```
I think it's fine to keep seed as a kwarg of the test. While we cannot afford to run the CI with with many different seed values, keeping the seed as a kwargs makes it easier to locally add a line such as 

```python
@pytest.mark.parametrize("seed", range(10))
```

locally to run the test with more seeds to check that it's not too sensitive to a specific choice of the seed value. Otherwise it's a sign that the test is brittle and can fail for no good reason on different platforms (Operation System / CPU architecture / user land bitness...).

lobpcg replied:
done

ogrisel reviewed (commented):
some more feedback:

ogrisel commented:
> [MRG] propagate eigen_tol #11968
> Simplify spectral clustering solver logic #14713

Also cross-linking with #15136 which is related to #14713.

lobpcg committed:
Update test_spectral_embedding.py

seed=36 goes back to args everywhere

lobpcg committed:
Update test_spectral_embedding.py

trying to fix pytest.raises(ValueError)

lobpcg committed:
Update test_spectral_embedding.py

black

lobpcg committed:
Update _spectral_embedding.py

randn back to rand . Will be changed in a separate PR

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -117,33 +125,59 @@ def test_spectral_embedding_two_components(seed=36):
-    label_ = np.array(embedded_coordinate.ravel() < 0, dtype="float")
-    assert normalized_mutual_info_score(true_label, label_) == pytest.approx(1.0)
+    for dtype in [np.float32, np.float64]:
+        if eigen_solver == "amg" and not amg_loaded:
```
I think that we can isolate the testing regarding the erroring of PyAMG not being available.
Like this, we could avoid the indentation with the for loop (with a parametrization) and skip if PyAMG is not installed.

glemaitre replied:
So we can add the following test:

```python
@pytest.mark.skipif(
    not is_pyamg_not_available,
    reason="PyAMG is installed and we should not test for an error."
)
def test_error_pyamg_not_available():
    se_precomp = SpectralEmbedding(
        n_components=2,
        affinity="rbf",
        eigen_solver="amg",
    )
    err_msg = "The eigen_solver was set to 'amg', but pyamg is not available."
    with pytest.raises(ValueError, match=err_msg):
        se_precomp.fit_transform(S)
```

lobpcg replied:
@glemaitre My check also tested that there is no error if PyAMG is installed. 
Anyway, it's not related to this PR - maybe should be done in a different PR?

glemaitre replied:
> My check also tested that there is no error if PyAMG is installed.

Basically, we already have other tests that check some normal behaviour when `amg` is installed. So there is no need to do anything when `amg` is installed in this test.

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -19,6 +19,13 @@
 from sklearn.utils._testing import assert_array_almost_equal
 from sklearn.utils._testing import assert_array_equal
 
+try:
+    from pyamg import smoothed_aggregation_solver  # noqa
+
+    amg_loaded = True
+except ImportError:
+    amg_loaded = False
+
```
To skip the test that does if `PyAMG` is not installed, we will use `pytest`

```suggestion
try:
    import pyamg
    is_pyamg_not_available = False
except ImportError:
    is_pyamg_not_available = True
skip_if_no_pyamg = pytest.mark.skipif(
    is_pyamg_not_available, reason="PyAMG is required for the tests in this function."
)
```

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -85,7 +92,8 @@ def test_sparse_graph_connected_component():
 
 
-def test_spectral_embedding_two_components(seed=36):
+@pytest.mark.parametrize("eigen_solver", ("arpack", "lobpcg", "amg"))
```
Now, we can only focus on the real purpose of the test and discard the `PyAMG` issue.

```suggestion
@pytest.mark.parametrize(
    "eigen_solver",
    [
        "arpack",
        "lobpcg",
        pytest.param("amg", marks=skip_if_no_pyamg),
    ]
)
@pytest.mark.parametrize("dtype", [np.float32, np.float64])
```

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -85,7 +92,8 @@ def test_sparse_graph_connected_component():
 
-def test_spectral_embedding_two_components(seed=36):
+@pytest.mark.parametrize("eigen_solver", ("arpack", "lobpcg", "amg"))
+def test_spectral_embedding_two_components(eigen_solver, seed=36):
```
```suggestion
def test_spectral_embedding_two_components(eigen_solver, dtype, seed=36):
```

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -117,33 +125,59 @@ def test_spectral_embedding_two_components(seed=36):
+        else:
+            embedded_coordinate = se_precomp.fit_transform(affinity.astype(dtype))
+            # thresholding on the first components using 0.
+            label_ = np.array(embedded_coordinate.ravel() < 0, dtype="float")
```
```suggestion
            label_ = np.array(embedded_coordinate.ravel() < 0, dtype=np.float64)
```

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -117,33 +125,59 @@ def test_spectral_embedding_two_components(seed=36):
 
 @pytest.mark.parametrize("X", [S, sparse.csr_matrix(S)], ids=["dense", "sparse"])
-def test_spectral_embedding_precomputed_affinity(X, seed=36):
+@pytest.mark.parametrize("eigen_solver", ("arpack", "lobpcg", "amg"))
```
we can make the same trick with `skip_if_no_pyamg`

lobpcg commented:
All requests have been addressed. Ready to merge?

glemaitre reviewed (commented):
I made the first pass to improve the PyAMG handling. I will check closer the numerical side now.

lobpcg committed:
Apply suggestions from code review

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -117,33 +134,59 @@ def test_spectral_embedding_two_components(seed=36):
+            label_ = np.array(embedded_coordinate.ravel() < 0, dtype=np.float64)
+            assert normalized_mutual_info_score(true_label, label_) == pytest.approx(
+                1.0
+            )
```
When passing `np.float32`, we should ensure that the fitted attributes (`embedding_` and `affinity_matrix_`) are resulting in a matrix with the same `dtype`. I assume that we want the same behaviour for the data returned by `fit_transform`. So we need to assert the `dtype` depending on the input `dtype`.

lobpcg replied:
the labels are integers, so on 32-bit OS they will be int32 no matter what the dtype is for the data

glemaitre replied:
The issue here is more about the old code using `dtype="float"` that will not be the same dtype depending on the platform. `np.int32` or `np.int64` would also be fine.

lobpcg replied:
@glemaitre could you please propose a specific commit to avoid any misunderstanding?

lobpcg committed:
Update test_spectral_embedding.py

complete removing amg_loaded

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -205,7 +248,8 @@ def test_spectral_embedding_callable_affinity(X, seed=36):
     "ignore:scipy.linalg.pinv2 is deprecated:DeprecationWarning:pyamg.*"
 )
-def test_spectral_embedding_amg_solver(seed=36):
+@pytest.mark.parametrize("dtype", (np.float32, np.float64))
```
since we have our created our marker, let's use it here.

```suggestion
@pytest.mark.skipif(
    is_pyamg_not_available, reason="PyAMG is not installed and thus we cannot test."
)
@pytest.mark.parametrize("dtype", (np.float32, np.float64))
```

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -205,7 +248,8 @@ def test_spectral_embedding_callable_affinity(X, seed=36):
+@pytest.mark.parametrize("dtype", (np.float32, np.float64))
+def test_spectral_embedding_amg_solver(dtype, seed=36):
     # Test spectral embedding with amg solver
     pytest.importorskip("pyamg")
```
we can remove then this line

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -258,12 +302,13 @@ def test_spectral_embedding_amg_solver(seed=36):
+@pytest.mark.parametrize("dtype", (np.float32, np.float64))
+def test_spectral_embedding_amg_solver_failure(dtype, seed=36):
     # Non-regression test for amg solver failure (issue #13393 on github)
     pytest.importorskip("pyamg")
```
We can as well use the `skipif` here

glemaitre commented:
Regarding the assertion of the output and fitted attributes data type, we could indeed create a specific function `test_spectral_embeddding_preserving_dype` parameterized with the float64 and float32 types that only check this particular behaviour. It will not clutter the semantic of the other tests.

lobpcg committed:
Update sklearn/manifold/tests/test_spectral_embedding.py

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

glemaitre commented on doc/whats_new/v1.1.rst:
```diff
@@ -117,6 +117,12 @@ Changelog
+:mod:`sklearn.manifold`
+.......................
+
+- |Enhancement| :func:`manifold.spectral_embedding` supports `np.float32` dtype.
+  :pr:`21534` by :user:`Andrew Knyazev <lobpcg>`.
```
```suggestion
- |Enhancement| :func:`manifold.spectral_embedding` and
   :class:`manifold.SpectralEmbedding` supports `np.float32` dtype and will
   preserve this dtype.
  :pr:`21534` by :user:`Andrew Knyazev <lobpcg>`.
```

lobpcg committed:
Update doc/whats_new/v1.1.rst

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

lobpcg committed:
Update test_spectral_embedding.py

`import pyamg` fails lint

lobpcg committed:
Update test_spectral_embedding.py

black

lobpcg commented:
> Regarding the assertion of the output and fitted attributes data type, we could indeed create a specific function `test_spectral_embeddding_preserving_dype` parameterized with the float64 and float32 types that only check this particular behaviour. It will not clutter the semantic of the other tests.

@glemaitre could you please propose a specific commit?

lobpcg committed:
Update test_spectral_embedding.py

black

lobpcg committed:
Update test_spectral_embedding.py

lint

lobpcg committed:
Update test_spectral_embedding.py

@pytest.mark.skipif(
    is_pyamg_not_available, reason="PyAMG is not installed and thus we cannot test."
)
replaces pytest.importorskip("pyamg")

glemaitre commented:
This would be the test that I have in mind:

```python
@pytest.mark.parametrize(
    "eigen_solver",
    [
        "arpack",
        "lobpcg",
        pytest.param("amg", marks=skip_if_no_pyamg),
    ],
)
@pytest.mark.parametrize("dtype", [np.float32, np.float64])
def test_spectral_embedding_preserves_dtype(eigen_solver, dtype):
    """Check that `SpectralEmbedding is preserving the dtype of the fitted
    attribute and transformed data.

    Ideally, this test should be covered by the common test
    `check_transformer_preserve_dtypes`. However, this test only run
    with transformers implementing `transform` while `SpectralEmbedding`
    implements only `fit_transform`.
    """
    X = S.astype(dtype)
    se = SpectralEmbedding(
        n_components=2, affinity="rbf", eigen_solver=eigen_solver, random_state=0
    )
    X_trans = se.fit_transform(X)

    assert X_trans.dtype == dtype
    assert se.embedding_.dtype == dtype
    assert se.affinity_matrix_.dtype == dtype
```

glemaitre commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -117,31 +135,46 @@ def test_spectral_embedding_two_components(seed=36):
+    for dtype in [np.float32, np.float64]:
+        embedded_coordinate = se_precomp.fit_transform(affinity.astype(dtype))
+        # thresholding on the first components using 0.
+        label_ = np.array(embedded_coordinate.ravel() < 0, dtype=np.float64)
```
If we expect the label to be an integer, so the following will be fine.

```suggestion
        label_ = np.array(embedded_coordinate.ravel() < 0, dtype=np.int64)
```

lobpcg replied:
would this work in 32-bit OS? I have tried  similar changes suggested earlier by @ogrisel and found that tests fail in 32-bit OS

glemaitre replied:
I don't think that with integer we should have an issue here.

lobpcg commented:
@glemaitre I'll try adding your new tests just above and https://github.com/scikit-learn/scikit-learn/pull/21534#discussion_r743022081 after getting all the tests pass again with the new changes

glemaitre commented on doc/whats_new/v1.1.rst:
```diff
@@ -117,6 +117,14 @@ Changelog
+:mod:`sklearn.manifold`
+.......................
+
+- |Enhancement| :func:`manifold.spectral_embedding` and
+   :class:`manifold.SpectralEmbedding` supports `np.float32` dtype and will
+   preserve this dtype.
+  :pr:`21534` by :user:`Andrew Knyazev <lobpcg>`.
```
The CI fails because of this one (that I probably introduced in my suggestion :))

```suggestion
- |Enhancement| :func:`manifold.spectral_embedding` and
  :class:`manifold.SpectralEmbedding` supports `np.float32` dtype and will
  preserve this dtype.
  :pr:`21534` by :user:`Andrew Knyazev <lobpcg>`.
```

lobpcg replied:
yes, just fixed it already

lobpcg committed:
Update v1.1.rst

typos

lobpcg committed:
Update test_spectral_embedding.py

2 new tests proposed by @glemaitre

lobpcg committed:
Update sklearn/manifold/tests/test_spectral_embedding.py

int64 for labels

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

lobpcg commented:
@glemaitre both your new tests pass. Does one of them cover already your suggestion https://github.com/scikit-learn/scikit-learn/pull/21534#discussion_r743059485 ? This is the only one remains open as far as I can see?

ogrisel commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -19,6 +19,15 @@
 from sklearn.utils._testing import assert_array_almost_equal
 from sklearn.utils._testing import assert_array_equal
 
+try:
+    from pyamg import smoothed_aggregation_solver  # noqa
+
+    is_pyamg_not_available = False
+except ImportError:
+    is_pyamg_not_available = True
+skip_if_no_pyamg = pytest.mark.skipif(
+    is_pyamg_not_available, reason="PyAMG is required for the tests in this function."
+)
```
I would prefer to avoid reasoning with multiple negations:

```suggestion
try:
    from pyamg import smoothed_aggregation_solver  # noqa

    pyamg_available = True
except ImportError:
    pyamg_available = False
skip_if_no_pyamg = pytest.mark.skipif(
    not pyamg_available, reason="PyAMG is required for the tests in this function."
)
```

the tests will need to be adapted though.

lobpcg replied:
edited directly together with tests

ogrisel reviewed (approved):
One final nitpick but otherwise LGTM! Thank you very much for your perseverance :)

lobpcg committed:
Update test_spectral_embedding.py

https://github.com/scikit-learn/scikit-learn/pull/21534#discussion_r743553387

lobpcg committed:
Update test_spectral_embedding.py

typo

lobpcg committed:
Update test_spectral_embedding.py

black

lobpcg commented:
@ogrisel thanks for reviewing promptly! "Perseverance" is my nick name.  Life is too short for our coding efforts to be wasted :-)

jjerphan commented on sklearn/manifold/_spectral_embedding.py:
```diff
@@ -366,6 +369,7 @@ def spectral_embedding(
             # approximation X to eigenvectors
             X = random_state.rand(laplacian.shape[0], n_components + 1)
             X[:, 0] = dd.ravel()
+            X = X.astype(laplacian.dtype)
```
I wish it were possible to specify dtype when using numpy PRNG!

glemaitre replied:
I thought the same. But we need to discuss the SLEP about the random state

jjerphan commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -19,6 +19,15 @@
 from sklearn.utils._testing import assert_array_almost_equal
 from sklearn.utils._testing import assert_array_equal
 
+try:
+    from pyamg import smoothed_aggregation_solver  # noqa
+
+    pyamg_available = True
+except ImportError:
+    pyamg_available = False
+skip_if_no_pyamg = pytest.mark.skipif(
+    not pyamg_available, reason="PyAMG is required for the tests in this function."
+)
```
Nitpick.
```suggestion
try:
    from pyamg import smoothed_aggregation_solver  # noqa
    pyamg_available = True
except ImportError:
    pyamg_available = False

skip_if_no_pyamg = pytest.mark.skipif(
    not pyamg_available, reason="PyAMG is required for the tests in this function."
)
```

lobpcg replied:
black complained about not have an empty line after import

jjerphan commented on sklearn/manifold/tests/test_spectral_embedding.py:
```diff
@@ -117,31 +135,46 @@ def test_spectral_embedding_two_components(seed=36):
-    # thresholding on the first components using 0.
-    label_ = np.array(embedded_coordinate.ravel() < 0, dtype="float")
-    assert normalized_mutual_info_score(true_label, label_) == pytest.approx(1.0)
+    for dtype in [np.float32, np.float64]:
+        embedded_coordinate = se_precomp.fit_transform(affinity.astype(dtype))
+        # thresholding on the first components using 0.
+        label_ = np.array(embedded_coordinate.ravel() < 0, dtype=np.int64)
+        assert normalized_mutual_info_score(true_label, label_) == pytest.approx(1.0)
```
Isn't it the goal of the parametrization on `dtype`?
```suggestion
    embedded_coordinate = se_precomp.fit_transform(affinity.astype(dtype))
    # thresholding on the first components using 0.
    label_ = np.array(embedded_coordinate.ravel() < 0, dtype=np.int64)
    assert normalized_mutual_info_score(true_label, label_) == pytest.approx(1.0)
```

glemaitre replied:
Whoops. yes indeed.

lobpcg replied:
To my recollection, this specific test is not not parametrized, so it's done differently, following an example from test_cluster_qr in
https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/cluster/tests/test_spectral.py

Which also BTW has given me the way of treating pyamg
```
try:
    from pyamg import smoothed_aggregation_solver  # noqa

    amg_loaded = True
except ImportError:
    amg_loaded = False
```
Maybe someone want to now edit https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/cluster/tests/test_spectral.py making it consistent with this PR.

lobpcg replied:
Anyway, please feel free to edit anything you like, of course. My only interest was to add the float32 support, rather than to beautify the tests.

jjerphan replied:
I just have opened https://github.com/scikit-learn/scikit-learn/pull/21564 accordingly.

glemaitre reviewed (approved):
LGTM

glemaitre commented:
Thanks @lobpcg all good for me.

lobpcg commented:
@glemaitre @ogrisel  - thanks for your help!

jjerphan reviewed (commented):
Well, I was writing this review. :smile:

lobpcg commented:
@jjerphan - Thanks for taking over!
